### DISCUSSION TITLE
Standard interface for alternative backends (compilation targets)

### DISCUSSION THREAD

**Discussion by SichangHe** [gleam-lang/gleam#3352] [Ideas & suggestions]
Gleam currently compiles to Erlang or JavaScript, officially, but there is [an unofficial Nix backend that forks the whole compiler](https://github.com/Glistix/glistix?tab=readme-ov-file), and there are [interests for C/C++ targets](https://github.com/gleam-lang/gleam/discussions/2662), [Go Lang](https://github.com/gleam-lang/gleam/discussions/1736), and [WASM target](https://github.com/gleam-lang/gleam/issues/109).

It would be nice if the implementation of a compilation backend can be separated from the rest of the compiler, so these efforts can be parallelized and the Gleam project does not necessarily have to maintain all of them.

Hope this helps,\
Steven Hé (Sīchàng)

**Comment by lpil** [comment]
Making compiler public compiler APIs would be a much larger maintenance burden than the two existing backends. We have little developer time and very little money to spend here, so unless we get significantly more funding or complete all higher impact work then this is not something that we can do.

**Reply by SichangHe** [reply]
It obviously would be a lot if the backends are supposed to plug into something in the middle.

Instead, what about just having `gleam build` call these backends and let them do all the steps in the compilation? The main benefit would be to have these targets coexist on one's system and project.

**Reply by lpil** [reply]
What advantage would `gleam build --target c` have over `gleam-c build`?

**Reply by SichangHe** [reply]
Okay, I should be clearer here. It could be something like this:

1. The user specifies which backends to use in `gleam.toml`, along with the shared or specific configurations.
1. When the user builds, they use `gleam build`, and either builds the default targets or whatever they specify. The user does not need to know, for example, how to use `gleam-c`, so if A lets B try their codebase, B only needs to know they need to install this dependency called a Gleam C backend, and just run `gleam build` like for any other Gleam project.
1. This also applies to other Gleam commands, like `gleam test`. It would be nicer to just run `gleam test` to run tests for all targets in `gleam.toml`[^all-targets], instead of running `gleam test && glistix test && gleam-c test`.

[^all-targets]: This is related to https://github.com/gleam-lang/gleam/discussions/1638.

**Reply by lpil** [reply]
Ah! So there's two features:

1. A way to run commands for multiple targets.
2. Some support for other backends.

From my point of view we are extremely low on funding so anything that produces work without clear advantage to all Gleam users isn't something we can consider adding to the backlog. This makes anything relating to additional compiler backends unviable.

**Reply by SichangHe** [reply]
> * A way to run commands for multiple targets.

This is planned, right?

> * Some support for other backends.

I think the question is: what is a good way for one to implement a standalone alternative Gleam backend and make it work with the rest of the ecosystem? Glistix forked the whole compiler, which does not seem the nicest to me; maybe we should ask @PgBiel. But, the fact that Glistix works indicates that other "unofficial" Gleam backends could be a common thing, and the Gleam team does not maintain them anyway.

**Reply by lpil** [reply]
No plans for that currently. If you have ideas make a discussion to detail the use cases, prior art, and what the design might be.

The recommendation is to fork. As I said we are extremely low on funding and cannot afford to spend any time on supporting unofficial Gleam derivatives, and we also want to make very clear that they are unofficial. The more official something looks the more time we have to spend answering support requests which turn out to be about the third party project, and this is already a notable drain of core's limited developer time.

**Reply by SichangHe** [reply]
> > * A way to run commands for multiple targets.
> 
> This is planned, right?

> No plans for that currently. If you have ideas make a discussion to detail the use cases, prior art, and what the design might be.

Are you talking about the former quote or the support for backends?

> The more official something looks the more time we have to spend answering support requests which turn out to be about the third party project, and this is already a notable drain of core's limited developer time.

Okay, this is the real big issue here I overlooked. If B runs `gleam build` to build A's custom target and it does not work, B would likely seek answers from the Gleam project instead of that custom backend.

**Reply by lpil** [reply]
The former!

**Comment by SichangHe** [comment]
So, @lpil, the TL;DR seems to be: your take is that Gleam should only support backends that are part of the project itself, and unofficial backends should fork the compiler; having more official backends is desirable but a future work. Did I get it close?

**Reply by lpil** [reply]
Yes, we will not accept any additional workload. We do not have enough money.

**Comment by PgBiel** [comment]
Hi,

I'd like to provide my two cents on this.

Having Gleam support compiling to extra targets (invoking 3rd-party compilers or providing APIs), as well as upstreaming those in any way, has already been discussed, and it likely won't happen.

However, one particular pain point right now is that extra targets very much need an extra value for `@target(...)`, however invalid values for `@target` fail at the parser level. This means that a package author which wants to add support for Glistix users, for example, won't be able to do so without creating and publishing a copy of their package. This means they'd have two copies on Hex: one with only the official targets and one only with an additional (unofficial) target.

This makes me wonder if we can improve this situation. One solution I thought of: what if we allowed specifying an `extra-targets = ["nix"]` field (bikesheddable name) which would specify extra valid `@target` values, but those values would always be ignored by the official compiler? That way, one package can optionally have support for JS, Erlang and desired extra targets without affecting users of the official compiler. What do you think?

**Reply by lpil** [reply]
No changes will be accepted to core for other targets unless we had significantly more funding.
